### ISSUE TITLE
[Bug] Ember 4 regression with passed-in event handlers

### ISSUE THREAD

**Issue by boris-petrov** [emberjs/ember.js#19976]
### 🐞 Describe the Bug
I'm not sure the issue name is good but...

### 🔬 Minimal Reproduction
[This repo](https://github.com/boris-petrov/ember-4-bug). Clone, `npm install`, `ember serve`. Click the `click me!` text. Nothing happens (but it should). If you remove the last line in `app/templates/application.hbs` - it does happen. If you downgrade Ember's version to 3.28.x - it also does happen.

### 😕 Actual Behavior
Nothing changes.

### 🤔 Expected Behavior
To display `true` because of the `set` call.

### 🌍 Environment

- Ember: 4.1.0 (happens also with 4.2.0)
- Node.js/npm: 17
- OS: Linux
- Browser: Chrome 98

### ➕ Additional Context
[Discord link](https://discord.com/channels/480462759797063690/480462759797063692/943133829307199508)

**Comment by Timbuktu1982** [comment]
**Hi Boris,**

if you use the normal (actual) ember style for usage of components (on-modifier, <ComponentName /> instead of {{#componentName /}}, you have no problems... Alternatively you can use {{component ComponentName /}}..

Following changed code is functional:

`{{page-title 'Ember4Bug'}}

{{later (set this 'array' (array 1))}}

{{#each this.array as |unused|}}
  aaa -
  {{this.foobar}}
  - aaa

  <ClassicComponent {{on 'click' (set this 'foobar' true)}}>
    click me!
  </ClassicComponent>

{{/each}}

<ClassicComponent />`

**Comment by boris-petrov** [comment]
@Timbuktu1982 - yes, this is exactly how I fixed the issue in my own code. However, this is still a bug (which could be because of some bigger issue underneath).

**Comment by Timbuktu1982** [comment]
Ok. But i'am not understand, why you use the old style and not the new Style?

**Comment by Timbuktu1982** [comment]
Here the documentation for this:

https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rule/no-curly-component-invocation.md

**Comment by boris-petrov** [comment]
Because it was legacy code written years ago. :) Again - of course the "new" way is better (and it works) but still - that's a bug in Ember that would be nice to be fixed. :)

**Comment by Timbuktu1982** [comment]
Yes, I understand that up front, but I think it's necessary to touch and adjust all the code anyway. I know, in many cases it is many work.
As for the old style, I would have thought that the old style is generally no longer applicable.

**Comment by Timbuktu1982** [comment]
Ah ok, I saw in the documentation that both ways are still usable....

https://guides.emberjs.com/release/upgrading/current-edition/templates/

**Comment by kategengler** [comment]
We plan to deprecate @ember/component https://github.com/emberjs/rfcs/pull/1216 so this is unlikely to be fixed
